Userland/Libraries/LibGfx/ImageFormats/WebPLoader.cpp

```diff
@@ -297,6 +298,352 @@ static ErrorOr<VP8LHeader> decode_webp_chunk_VP8L_header(WebPLoadingContext& con
     return VP8LHeader { width, height, is_alpha_used };
 }
 
+// https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#61_overview
+// "From here on, we refer to this set as a prefix code group."
+class PrefixCodeGroup {
+public:
+    Compress::CanonicalCode& operator[](int i) { return m_codes[i]; }
+    Compress::CanonicalCode const& operator[](int i) const { return m_codes[i]; }
+
+private:
+    Array<Compress::CanonicalCode, 5> m_codes;
+};
+
+// https://developers.google.com/speed/webp/docs/webp_lossless_bitstream_specification#621_decoding_and_building_the_prefix_codes
+static ErrorOr<Compress::CanonicalCode> decode_webp_chunk_VP8L_prefix_code(WebPLoadingContext& context, LittleEndianInputBitStream& bit_stream, size_t alphabet_size)
+{
+    // prefix-code           =  simple-prefix-code / normal-prefix-code
+    bool is_simple_code_length_code = TRY(bit_stream.read_bits(1));
+    dbgln_if(WEBP_DEBUG, "is_simple_code_length_code {}", is_simple_code_length_code);
+
+    Vector<u8, 286> code_lengths;
+
+    if (is_simple_code_length_code) {
+        TRY(code_lengths.try_resize(alphabet_size));
+
+        int num_symbols = TRY(bit_stream.read_bits(1)) + 1;
+        int is_first_8bits = TRY(bit_stream.read_bits(1));
+        u8 symbol0 = TRY(bit_stream.read_bits(1 + 7 * is_first_8bits));
+        dbgln_if(WEBP_DEBUG, "  symbol0 {}", symbol0);
+
+        if (symbol0 >= code_lengths.size())
+            return Error::from_string_literal("symbol0 out of bounds");
+        code_lengths[symbol0] = 1;
+        if (num_symbols == 2) {
+            u8 symbol1 = TRY(bit_stream.read_bits(8));
+            dbgln_if(WEBP_DEBUG, "  symbol1 {}", symbol1);
+
+            if (symbol1 >= code_lengths.size())
+                return Error::from_string_literal("symbol1 out of bounds");
+            code_lengths[symbol1] = 1;
+        }
+
+        return Compress::CanonicalCode::from_bytes(code_lengths);
+    }
+
+    // This has plenty in common with deflate (cf DeflateDecompressor::decode_codes() in Deflate.cpp in LibCompress)
+    // Symbol 16 has different semantics, and kCodeLengthCodeOrder is different. Other than that, this is virtually deflate.
+    // (...but webp uses 5 different prefix codes, while deflate doesn't.)
+    int num_code_lengths = 4 + TRY(bit_stream.read_bits(4));
+    dbgln_if(WEBP_DEBUG, "  num_code_lengths {}", num_code_lengths);
+
+    // "If num_code_lengths is > 19, the bit_stream is invalid. [AMENDED3]"
+    if (num_code_lengths > 19)
+        return context.error("WebPImageDecoderPlugin: invalid num_code_lengths");
+
+    constexpr int kCodeLengthCodes = 19;
+    int kCodeLengthCodeOrder[kCodeLengthCodes] = { 17, 18, 0, 1, 2, 3, 4, 5, 16, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 };
```

Comment:
This could be an `Array<int, 19>`

Reply:
This is copied from the spec as-is. Should I still change it?

Reply:
Meh, if it's literally code that you copied from the spec I guess it makes sense to leave it as-is. Serenitification of code helps in other ways though so feel free to ignore or apply this suggestion as you see fit!

Reply:
Let's keep it as-is for now then :)